search/utils.py

```diff
@@ -74,26 +88,44 @@ def search_index(tokens=None, repo_slug=None, sort_by=None, terms=None):
         terms = {}
     if tokens is None:
         # Retrieve everything.
-        search = Search(index=INDEX_NAME, doc_type=DOC_TYPE).query()
+        search = Search(index=INDEX_NAME, doc_type=DOC_TYPE)
     else:
         # Search on title, description, and content_xml (minus markup).
         search = Search(index=INDEX_NAME, doc_type=DOC_TYPE)
         multi = query.MultiMatch(
             query=tokens, fields=["title", "description", "content_stripped"])
         search = search.query(multi)
-    if terms != {}:
-        # Filter further on taxonomy terms.
-        search = search.query("match", **terms)
+
+    # Filter further on taxonomy terms.
+    for key, value in terms.items():
+        search = search.query("match", **{key: value})
+
     if repo_slug is not None:
         # Filter further on repository.
         search = search.query("match", repository=repo_slug)
-    if sort_by is not None:
+    if sort_by is None:
+        # Always sort by ID to preserve ordering.
+        search = search.sort("id")
+    else:
         # Temporary workaround; the values in sorting.py should be updated,
         # but for now Haystack is still using them. Also, the hyphen is
         # required because we sort the numeric values high to low.
-        if not sort_by.startswith("title"):
-            sort_by = "-xa_{0}".format(sort_by)
+        if "title" not in sort_by:
+            reverse = sort_by.startswith("-")
+            if reverse:
+                sort_by = sort_by[1:]
+            if "xa" not in sort_by:
+                sort_by = "xa_{0}".format(sort_by)
+            if reverse:
+                sort_by = "-{0}".format(sort_by)
+        # Always sort by ID to preserve ordering.
```

Comment:
At present, this just takes the stuff passed into `construct_queryset`.

Once Haystack is gone, this bit can be stripped out and calls to `search_index` can be passed the appropriate arguments.